### PR TITLE
Reworked byte <-> LSP position conversions

### PR DESCRIPTION
**Pull request by AndrewF001**

### CODE REVIEW

File: crates/analyzer-host/src/lsp_impl/active_initialized.rs [hunk 1]
```diff
@@ -172,6 +173,7 @@ async fn on_text_document_did_change(
 	}
 
 	// TODO: avoid cloning
+	// implement LspPos::lazy_add
```
Comment (AndrewF001) [inline crates/analyzer-host/src/lsp_impl/active_initialized.rs:176]: Currently lazy_add is never used but this is where it would go. `analyzer.update()` is where the LspPos gets updated but that will bypass the use of `TextDocumentContentChangeEvent`. Wondering if the best way would be for `update()` to take a `Buffer` argument instead of file?

File: crates/analyzer-core/tests/lsp_position_struct.rs [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+
+					let (l1, c1) = (start.line, start.character);
+					let (l2, c2) = (end.line, end.character);
+					let event = change_event((l1, c1), (l2, c2), change.to_string());
+
+					let mut lsp = original_lsp.clone();
+
+					let clock = Instant::now();
+					lsp.lazy_add(&event);
+					lazy_timer += clock.elapsed();
```
Comment (AndrewF001) [inline crates/analyzer-core/tests/lsp_position_struct.rs:407]: For smaller files, `lazy_add()` is actually slower than just re-parsing the file. From tests, a file with 60 characters and 12 '\n' occurrences is the turning point for when `lazy_add()` is more efficient

**Review by AndrewF001 — COMMENTED** [review]
This PR has been a long time in the works but finally got the Lazy add to work and hope the tests prove that!

File: crates/analyzer-core/src/lsp_position_struct.rs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+		let eof = string.as_bytes().last() == Some(&b'\n');
+
+		(result, eof)
+	}
+
+	pub fn parse_file(file: &String) -> Self {
+		let parse = LspPos::parse_string(&file);
+		LspPos { ranges: parse.0, eof: parse.1 }
+	}
```
Comment (timjroberts) [inline crates/analyzer-core/src/lsp_position_struct.rs:36]: You can make use of the tuple type returned from `parse_string(...)` here to make the code more concise:

```rust
let (ranges, eof) = LspPos::parse_string(&file);
LspPos { ranges, eof }
```

File: crates/analyzer-core/tests/lsp_position_struct.rs
Comment (timjroberts) [inline crates/analyzer-core/tests/lsp_position_struct.rs:1]: Might be worth adding some tests that include extended unicode characters. Put some emojis into the input and see if the ranges are still as expected.

**Review by timjroberts — APPROVED** [review]
Looks good to me.

**Comment by viluon** [comment]
I have a WIP review of this; sorry for taking so long. I'll review the new changes and submit the review today.

File: crates/analyzer-core/src/lsp_file.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+		// update file
+		let range = start_byte..end_byte;
+		info!("replacing range {:?} of {:?} with {:?}", range, &self.file[range.clone()], &changes.text);
+		self.file.replace_range(range, &changes.text);
+
+		// remove old ranges and add new ranges
+		let len = additional_ranges.len();
+		let s = (start_line).min(range_size);
+		let e = (end_line + 1).min(range_size);
+		self.ranges.splice(s..e, additional_ranges); // used for performance benifits
```
Comment (viluon) [inline crates/analyzer-core/src/lsp_file.rs:220]:
```suggestion
		self.ranges.splice(s..e, additional_ranges); // used for performance benefits
```

File: crates/analyzer-core/src/lsp_file.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+		}
+
+		// parse input
+		let mut additional_ranges = LspFile::parse_string(&changes.text);
+		let addition_byte: i64 = additional_ranges.last().map_or(-1, |value| *value as i64);
+
+		// align additions to their placement in current file
+		let start_byte = self.lsp_to_byte(&start_pos);
+		let end_byte = self.lsp_to_byte(&end_pos_exc);
+		for elm in &mut additional_ranges {
```
Comment (viluon) [inline crates/analyzer-core/src/lsp_file.rs:164]: Does `elm` stand for "end of line mark?"

File: crates/analyzer-core/src/lsp_file.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+		let len = additional_ranges.len();
+		let s = (start_line).min(range_size);
+		let e = (end_line + 1).min(range_size);
+		self.ranges.splice(s..e, additional_ranges); // used for performance benifits
+
+		// realign tail of old ranges
+		let diff = (addition_byte + 1) - (end_byte as i64 - start_byte as i64);
+		for elm in self.ranges.iter_mut().skip(start_line + len) {
+			*elm = (*elm as i64 + diff) as usize;
+		}
```
Comment (viluon) [inline crates/analyzer-core/src/lsp_file.rs:226]: This just updates the offsets of line ends after the edit point, right?

File: crates/analyzer-core/src/preprocessor.rs [hunk 7]
```diff
@@ -572,7 +573,8 @@ mod test {
 		let mut pp = PreprocessorState::new(|path| FileId::new(&db, path.into()), |_| unreachable!());
 
 		let test_id = FileId::new(&db, "<test-code>.p4".into());
-		let input = Buffer::new(&db, s.into());
+		let lsp_file = LspFile::new(&s.to_string());
+		let input = Buffer::new(&db, lsp_file);
```
Comment (viluon) [inline crates/analyzer-core/src/preprocessor.rs:577]: Based on this usage, maybe it'd be better to present a unified API with a constructor for `Buffer` and hide the `LspFile` abstraction within it.

File: crates/analyzer-core/tests/lsp_file.rs
Comment (viluon) [inline crates/analyzer-core/tests/lsp_file.rs:1]: The tests here are quite repetitive, I think it would be better to specify them with a local macro or simply an array of pairs to check, something like
```rs
let cases = [
  ("\n", vec![0]),
  ("012\n456\n\n9", vec![3, 7, 8, 9]),
  ...
];

for (s, ranges) in cases {
  let str = s.to_string();
  let file = LspFile::new(&str);
  assert_eq!(*file.get_file_content(), str);
  assert_eq!(*file.get_ranges(), ranges);
  assert!(!file.get_eof());
}
```

Similarly for the other patterns.

**Review by viluon — CHANGES_REQUESTED** [review]
Requesting changes mainly because of the crate dependencies. Some minor comments throughout. Otherwise LGTM!

File: crates/analyzer-core/src/lsp_file.rs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+		if byte_pos > *self.ranges.last().unwrap_or(&0) {
+			return Position { line: self.ranges.len() as u32, character: 0 }; // return next position of last line
+		}
+
+		let line = self.ranges.binary_search(&byte_pos).unwrap_or_else(|x| x);
+
+		// calculate character position in byte offset
+		let mut byte_count = if line == 0 { 0 } else { self.ranges[line - 1] + 1 };
+		let slice = self.file.slice(byte_count..self.ranges[line]).unwrap_or("").chars();
```
[the remark this answers is not in the document]
Comment (AndrewF001) [reply crates/analyzer-core/src/lsp_file.rs:110]: We're sacrificing querying performance for reduced memory usage & quicker copying performance by using UTF-8

File: crates/analyzer-core/src/lsp_file.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+	}
+
+	pub fn get_file_content(&self) -> &String { &self.file }
+
+	pub fn get_ranges(&self) -> &Vec<usize> { &self.ranges }
+
+	pub fn get_eof(&self) -> bool { self.file.chars().last().unwrap_or_default() == '\n' }
+
+	// helper function
+	fn parse_string(string: &String) -> Vec<usize> {
```
[the remark this answers is not in the document]
Comment (viluon) [reply crates/analyzer-core/src/lsp_file.rs:21]: I agree with the private visibility, I just thought it could be moved out of the `impl` block and into the root scope of `lsp_file.rs`.

File: crates/analyzer-core/tests/lsp_file.rs
Comment (AndrewF001) [reply crates/analyzer-core/tests/lsp_file.rs:1]: Moved repetitive code into a function. Still not sure about it, if an `Assert!` fails you don't know which test it is on first glance (have to backtrack the call stack). I would also heavily avoid using any logic in unit tests as it has to be simple to follow with zero points of failure.

File: crates/analyzer-core/src/lsp_file.rs [hunk 10]
```diff
@@ -0,0 +1,228 @@
+use analyzer_abstractions::{
+	lsp_types::{self, Position, TextDocumentContentChangeEvent},
+	tracing::info,
+};
+use logos::Source; // for slice
+
+#[derive(Clone, Debug)]
+pub struct LspFile {
+	file: String,       // File content, use OsString?
```
[the remark this answers is not in the document]
Comment (viluon) [reply crates/analyzer-core/src/lsp_file.rs:6]: It's not really a fault of the OS -- the file contents are up to the user. If the user accidentally runs something like `cp switch-overview.png switch.p4` then we can't process a change to `switch.p4` as it's (likely) not valid UTF-8, and Rust strings can only represent UTF-8 text. We'll need to refactor this at a later point, but, as I said, it's out-of-scope for this PR, so don't worry about it.
